### PR TITLE
[WIP] ASRTTS

### PR DESCRIPTION
**Pull request by creatorscan**

### CODE REVIEW

File: egs/librispeech/asrtts/cmd.sh [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+# If jobs failed, your configuration might be wrong for your environment.
+#
+#
+# The official documentaion for run.pl, queue.pl, slurm.pl, and ssh.pl:
+#   "Parallelization in Kaldi": http://kaldi-asr.org/doc/queue.html
+# =========================================================~
+
+
+# Select the backend used by run.sh from "local", "sge", "slurm", or "ssh"
+cmd_backend='sge'
```
Comment (sw005320) [inline egs/librispeech/asrtts/cmd.sh]: can you make it `local`?

File: egs/librispeech/asrtts/conf/train_pytorch_tacotron2+spkemb.yaml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+postnet-layers: 5 # if set 0, no postnet is used
+postnet-chans: 512
+postnet-filts: 5
+use-speaker-embedding: true
+
+# attention related
+atype: location
+adim: 128
+aconv-chans: 32
+aconv-filts: 15      # resulting in filter-size = aconv-filts * 2 + 1
```
Comment (sw005320) [inline egs/librispeech/asrtts/conf/train_pytorch_tacotron2+spkemb.yaml]: We observed a significant performance degradation when we reduce the filter size? Is it OK for your case?

Comment (sw005320) [reply egs/librispeech/asrtts/conf/train_pytorch_tacotron2+spkemb.yaml]: Sorry, this is about ASR. Please ignore my comments.

File: egs/librispeech/asrtts/local/rand_datagen.sh [hunk 3]
```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
```
Comment (sw005320) [inline egs/librispeech/asrtts/local/rand_datagen.sh]: Can you add the copyright header and also explanations of this script?

File: egs/librispeech/asrtts/path.sh [hunk 4]
```diff
@@ -0,0 +1,38 @@
+MAIN_ROOT=$PWD/../../..
+KALDI_ROOT=$MAIN_ROOT/tools/kaldi
+SPNET_ROOT=$MAIN_ROOT/espnet
+BUT_CONDA=/homes/kazi/baskar/anaconda3/
```
Comment (sw005320) [inline egs/librispeech/asrtts/path.sh]: Please push the original (common) `path.sh`

File: espnet/asr/asrtts_utils.py [hunk 5]
```diff
@@ -0,0 +1,905 @@
...
+# io related
+import kaldi_io_py
+import matplotlib
+import numpy as np
+import torch
+matplotlib.use('Agg')
+
+
+# * -------------------- training iterator related -------------------- *
+def make_batchset_asr(data, batch_size, max_length_in, max_length_out,
```
Comment (ShigekiKarita) [inline espnet/asr/asrtts_utils.py]: Check `espnet.utils` not to redefine these functions. I'm afraid this makes us hard to maintain and test.
For example,

- make_batchset_xxx -> espnet.utils.training.batchfy
- LoadInputsAndOutputs -> espnet.utils.io_utils
- other -> espnet.asr.asr_utils, etc

https://github.com/espnet/espnet/tree/v.0.4.0/espnet/utils

File: espnet/asr/pytorch_backend/asrtts.py [hunk 6]
```diff
@@ -0,0 +1,699 @@
...
+    trainer.extend(extensions.PrintReport(
+        report_keys), trigger=(REPORT_INTERVAL, 'iteration'))
+
+    trainer.extend(extensions.ProgressBar(update_interval=REPORT_INTERVAL))
+
+    # Run the training
+    trainer.run()
+
+
+def recog(args):
```
Comment (ShigekiKarita) [inline espnet/asr/pytorch_backend/asrtts.py]: Is this function same to the one in ASR? If so, do `from espnet.asr.pytorch_backend.asr import recog` instead.

File: egs/librispeech/asrtts/conf/train_pytorch_tacotron2+spkemb.yaml (hunk 2 above)
Comment (ShigekiKarita) [reply egs/librispeech/asrtts/conf/train_pytorch_tacotron2+spkemb.yaml]: ASR should be configured in YAML too.

**Comment by ShigekiKarita** [comment]
Wonderful! The diff is clearly tractable now.

File: egs/librispeech/asrtts/path.sh [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    source $MAIN_ROOT/tools/venv/etc/profile.d/conda.sh && conda deactivate && conda activate
+else
+    source $MAIN_ROOT/tools/venv/bin/activate
+fi
+export PATH=$MAIN_ROOT/utils:$MAIN_ROOT/espnet/bin:$PATH
+
+export OMP_NUM_THREADS=1
+
+# check extra module installation
+if ! which spm_decode > /dev/null; then
```
Comment (sw005320) [inline egs/librispeech/asrtts/path.sh]: Do you need this?
(maybe not, right?)

File: egs/librispeech/asrtts/run.sh [hunk 8]
```diff
@@ -0,0 +1,596 @@
...
+        asrttsexpdir=${asrttsexpdir}_rnnlmloss_${rnnlm_loss}
+        train_opts="$train_opts --rnnlm $rnnlm_model --rnnlm-conf $rnnlm_model_conf --rnnloss mmd" 
+    elif [ "$rnnlm_loss" = "kl" ]; then
+        asrttsexpdir=${asrttsexpdir}_rnnlmloss_${rnnlm_loss}
+        train_opts="$train_opts --rnnlm $rnnlm_model --rnnlm-conf $rnnlm_model_conf --rnnloss kl" 
+    fi
+    if [ $asrtts_train == 'true' ]; then
+        ${cuda_cmd} --gpu 1 ${asrttsexpdir}/train.log \
+        asrtts_train.py \
+        --ngpu 6 \
```
Comment (sw005320) [inline egs/librispeech/asrtts/run.sh]: `--ngpu ${ngpu}`?

File: egs/librispeech/asrtts/run.sh [hunk 9]
```diff
@@ -0,0 +1,596 @@
...
+        --outdir ${asrttsexpdir}/results \
+        --debugmode ${debugmode} \
+        --dict ${dict} \
+        --debugdir ${asrttsexpdir} \
+        --minibatches ${N} \
+        --verbose ${verbose} \
+        --resume ${resume} \
+        --train-json ${tr_json_list} \
+        --valid-json ${feat_dt_dir}/data.json \
+        --etype ${etype} \
```
Comment (sw005320) [inline egs/librispeech/asrtts/run.sh]: It is possible to use config for asrtts as well?
This makes your `run.sh` clean.

File: espnet/asr/asrtts_utils.py [hunk 10]
```diff
@@ -0,0 +1,255 @@
+#!/usr/bin/env python
+
+# Copyright 2017 Johns Hopkins University (Shinji Watanabe)
```
Comment (sw005320) [inline espnet/asr/asrtts_utils.py]: This is your work.
Update it to your name.

File: egs/librispeech/asrtts/run.sh [hunk 11]
```diff
@@ -0,0 +1,544 @@
...
+    vaddir=mfcc
+    for name in ${train_set} ${dev_set} ${eval_set}; do
+        if [ ! -s data/${name}_mfcc/feats.scp ]; then
+        utils/copy_data_dir.sh data/${name} data/${name}_mfcc
+        steps/make_mfcc.sh \
+            --mfcc-config conf/mfcc.conf \
+            --nj ${nj} --cmd "$train_cmd" \
+            data/${name}_mfcc exp/make_mfcc ${mfccdir}
+        utils/fix_data_dir.sh data/${name}_mfcc
+        sid/compute_vad_decision.sh --nj ${nj} --cmd "$train_cmd" \
```
Comment (potato-inoue) [inline egs/librispeech/asrtts/run.sh]: Where is this script (sid/compute_vad_decision.sh) ?

File: egs/librispeech/asrtts/run.sh [hunk 12]
```diff
@@ -0,0 +1,544 @@
...
+    if [ ! -e ${nnet_dir} ];then
+        echo "X-vector model does not exist. Download pre-trained model."
+        wget http://kaldi-asr.org/models/8/0008_sitw_v2_1a.tar.gz
+        tar xvf 0008_sitw_v2_1a.tar.gz
+        mv 0008_sitw_v2_1a/exp/xvector_nnet_1a exp
+        rm -rf 0008_sitw_v2_1a.tar.gz 0008_sitw_v2_1a
+    fi
+    # Extract x-vector
+    for name in ${train_set} ${dev_set} ${eval_set}; do
+        sid/nnet3/xvector/extract_xvectors.sh --cmd "$train_cmd --mem 4G" --nj ${nj} \
```
Comment (potato-inoue) [inline egs/librispeech/asrtts/run.sh]: Where is this script (sid/nnet3/xvector/extract_xvectors.sh) ?

File: egs/librispeech/asrtts/run.sh [hunk 13]
```diff
@@ -0,0 +1,544 @@
...
+    data2json.sh --feat ${feat_dt_dir}/feats.scp \
+         data/${dev_set} ${dict} > ${feat_dt_dir}/data.json
+    data2json.sh --feat ${feat_ev_dir}/feats.scp \
+         data/${eval_set} ${dict} > ${feat_ev_dir}/data.json
+    fi
+    # Make MFCCs and compute the energy-based VAD for each dataset
+    mfccdir=mfcc
+    vaddir=mfcc
+    for name in ${train_set} ${dev_set} ${eval_set}; do
+        if [ ! -s data/${name}_mfcc/feats.scp ]; then
```
Comment (potato-inoue) [inline egs/librispeech/asrtts/run.sh]: I think feats.scp is made by make_mfcc.sh.
This sentence will be skipped at re-running if the problem happen after make_mfcc.sh.
Is this all right?

File: egs/librispeech/asrtts/run.sh [hunk 14]
```diff
@@ -0,0 +1,544 @@
...
+        utils/copy_data_dir.sh data/${name} data/${name}_mfcc
+        steps/make_mfcc.sh \
```
Comment (potato-inoue) [inline egs/librispeech/asrtts/run.sh]: I think it's better to add option `--write-utt2dur false \` for avoiding below error.

```
Command line was: compute-mfcc-feats --write-utt2dur=ark,t:exp/make_mfcc/utt2dur.1 --verbose=2 --config=conf/mfcc.conf scp,p:exp/make_mfcc/wav_train_960_mfcc.1.scp ark:-
ERROR (compute-mfcc-feats[5.5.293~1-f8021]:Read():parse-options.cc:365) Invalid option --write-utt2dur=ark,t:exp/make_mfcc/utt2dur.1
```

File: egs/librispeech/asrtts/run.sh (hunk 11 above)
Comment (sw005320) [reply egs/librispeech/asrtts/run.sh]: @creatorscan, you need to add the symlink of the Kaldi sid directroy like  https://github.com/espnet/espnet/blob/master/egs/libritts/tts1/sid

**Comment by potato-inoue** [comment]
@creatorscan 
The script (data2json_asrtts.sh) is not found.
Could you add this script?

```
stage 5: ASR-TTS training, decode and synthesize
utils/copy_data_dir.sh: copied data from data/train_clean_360 to data/train_clean_360_rnd
utils/validate_data_dir.sh: Successfully validated data-directory data/train_clean_360_rnd
local/rand_datagen.sh: line 58: data2json_asrtts.sh: command not found
```

**Comment by creatorscan** [comment]
> @creatorscan
> The script (data2json_asrtts.sh) is not found.
> Could you add this script?
> 
> ```
> stage 5: ASR-TTS training, decode and synthesize
> utils/copy_data_dir.sh: copied data from data/train_clean_360 to data/train_clean_360_rnd
> utils/validate_data_dir.sh: Successfully validated data-directory data/train_clean_360_rnd
> local/rand_datagen.sh: line 58: data2json_asrtts.sh: command not found
> ```

fixed the script. .

File: espnet/bin/asrtts_train.py [hunk 15]
```diff
@@ -0,0 +1,439 @@
...
+                        help='Output directory for debugging')
+    parser.add_argument('--resume', '-r', default='', nargs='?',
+                        help='Resume the training from snapshot')
+    parser.add_argument('--minibatches', '-N', type=int, default='-1',
+                        help='Process only N minibatches (for debug)')
+    parser.add_argument('--verbose', '-V', default=0, type=int,
+                        help='Verbose option')
+    parser.add_argument('--tensorboard-dir', default=None, type=str, nargs='?', help="Tensorboard log dir path")
+    # task related
+    parser.add_argument('--train-json', type=str, default=None, nargs='+',
```
Comment (potato-inoue) [inline espnet/bin/asrtts_train.py]: @creatorscan 
Did you make codes for using `args.train_json` of list type?

Error abut data-type was happened.
In ASR: `args.train_json` may be str.
In ASRTTS: `args.train_json` is list.
```
# asrtts_train.py --config conf/train_asrtts.yaml --ngpu 1 --backend pytorch --outdir exp/asrtts__exploss_pgrad/results --debugmode 1 --dict data/lang_char/train_960_units.txt --debugdir exp/asrtts__exploss_pgrad --minibatches 0 --verbose 0 --resume --train-json dump/train_clean_360/data_rnd.json dump/train_clean_100/data.json --valid-json dump/dev/data.json --opt adadelta --asr-model-conf exp/asr_/results/model.json --asr-model exp/asr_/results/model.acc.best --tts-model-conf exp/tts_/results/model.json --tts-model exp/tts_/results/model.loss.best --update-asr-only --policy-gradient 
# Started at Thu Sep 19 04:36:14 JST 2019
#
2019-09-19 04:36:15,845 (asrtts_train:376) WARNING: Skip DEBUG/INFO messages
2019-09-19 04:36:15,855 (asrtts_train:394) WARNING: CUDA_VISIBLE_DEVICES is not set.
2019-09-19 04:36:21,007 (e2e_asr:90) WARNING: Subsampling is not performed for vgg*. It is performed in max pooling layers at CNN.
Traceback (most recent call last):
  File "/work/abelab/k_inoue/debug/espnet/egs/librispeech/asrtts/../../../espnet/bin/asrtts_train.py", line 439, in <module>
    main(sys.argv[1:])
  File "/work/abelab/k_inoue/debug/espnet/egs/librispeech/asrtts/../../../espnet/bin/asrtts_train.py", line 427, in main
    train(args)
  File "/work/abelab/k_inoue/debug/espnet/espnet/asr/pytorch_backend/asr.py", line 356, in train
    with open(args.train_json, 'rb') as f:
TypeError: expected str, bytes or os.PathLike object, not list
# Accounting: time=12 threads=1
# Ended (code 1) at Thu Sep 19 04:36:26 JST 2019, elapsed time 12 seconds
```

**Pull request closed without merging**